string.c

```diff
@@ -3750,20 +3770,37 @@ rb_str_rindex(VALUE str, VALUE sub, long pos)
 
 /*
  *  call-seq:
- *     str.rindex(substring [, integer])   -> integer or nil
- *     str.rindex(regexp [, integer])   -> integer or nil
- *
- *  Returns the index of the last occurrence of the given <i>substring</i> or
- *  pattern (<i>regexp</i>) in <i>str</i>. Returns <code>nil</code> if not
- *  found. If the second parameter is present, it specifies the position in the
- *  string to end the search---characters beyond this point will not be
- *  considered.
- *
- *     "hello".rindex('e')             #=> 1
- *     "hello".rindex('l')             #=> 3
- *     "hello".rindex('a')             #=> nil
- *     "hello".rindex(?e)              #=> 1
- *     "hello".rindex(/[aeiou]/, -2)   #=> 1
+ *    string.rindex(substring, offset = self.length) -> integer or nil
+ *    string.rindex(regexp, offset = self.length) -> integer or nil
+ *
+ *  Returns the \Integer index of the _last_ occurrence of the given +substring+,
+ *  or +nil+ if none found:
+ *    'foo'.rindex('f') # => 0
+ *    'foo'.rindex('o') # => 2
+ *    'foo'.rindex('oo') # => 1
+ *    'foo'.rindex('ooo') # => nil
+ *
+ *  Returns the \Integer index of the _last_ match for the given \Regexp +regexp+,
+ *  or +nil+ if none found:
+ *    'foo'.rindex(/f/) # => 0
+ *    'foo'.rindex(/o/) # => 2
+ *    'foo'.rindex(/oo/) # => 1
+ *    'foo'.rindex(/ooo/) # => nil
+ *
+ *  \Integer argument +offset+, if given, specifies the position in the
+ *  string to _end_ the search:
```

Comment:
To me, this sounds not entirely correct (not your fault, I think it was wrong before).  The way it is worded, it sounds like `'foo'.rindex('oo', 1)` should return `nil`, since the search should end at the first `o` (offset 1) and not consider the second `o` (offset 2).

Admittedly, it is difficult to describe the precise behavior succinctly.  In practice, a non-negative offset specifies the maximum return value, and a negative offset specifies the maximum return value is (length + offset).  Maybe we can say:

```
*  \Integer argument +offset+, if given and non-negative, specifies the maximum starting position in the
*  string to _end_ the search:
```

```
*  If +offset+ is a negative \Integer, the maximum starting position in the
*  string to _end_ the search is the sum of the string's length and +offset+:
```

There may well be a better way to describe the precise behavior.

Reply:
Fixed.